pkg/fleets/controller.go

```diff
@@ -309,16 +314,102 @@ func (c *Controller) deleteEmptyGameServerSets(fleet *stablev1alpha1.Fleet, list
+		}
+	}
+
+	return fleet.LowerBoundReplicas(fleet.Spec.Replicas - stablev1alpha1.SumStatusAllocatedReplicas(rest)), nil
+}
+
+// rollingUpdateDeployment will do the rolling update of the old GameServers
+// through to the new ones, based on the fleet.Spec.Strategy.RollingUpdate configuration
+// and return the replica count for the active GameServerSet
+func (c *Controller) rollingUpdateDeployment(fleet *stablev1alpha1.Fleet, active *stablev1alpha1.GameServerSet, rest []*stablev1alpha1.GameServerSet) (int32, error) {
+	replicas, err := c.rollingUpdateActive(fleet, active, rest)
+	if err != nil {
+		return replicas, err
+	}
+	if err := c.rollingUpdateRest(fleet, rest); err != nil {
+		return replicas, err
+	}
+	return replicas, nil
+}
+
+// rollingUpdateActive applies the rolling update to the active GameServerSet
+// and returns what its replica value should be set to
+func (c *Controller) rollingUpdateActive(fleet *stablev1alpha1.Fleet, active *stablev1alpha1.GameServerSet, rest []*stablev1alpha1.GameServerSet) (int32, error) {
+	replicas := active.Spec.Replicas
+
+	if active.Spec.Replicas < fleet.Spec.Replicas && active.Spec.Replicas == active.Status.Replicas {
+		r, err := intstr.GetValueFromIntOrPercent(fleet.Spec.Strategy.RollingUpdate.MaxSurge, int(fleet.Spec.Replicas), true)
+		if err != nil {
+			return replicas, errors.Wrapf(err, "error calculating scaling gameserverset: %s", fleet.ObjectMeta.Name)
+		}
+		surge := int32(r)
+
+		// make sure we don't end up with more than the configured max surge
+		maxSurge := surge + fleet.Spec.Replicas
+		replicas = fleet.UpperBoundReplicas(replicas + surge)
+		total := stablev1alpha1.SumStatusReplicas(rest) + replicas
+		if total > maxSurge {
+			replicas = fleet.LowerBoundReplicas(replicas - (total - maxSurge))
+		}
+
+		// always leave room for Allocated GameServers
+		sumAllocated := stablev1alpha1.SumStatusAllocatedReplicas(rest)
+
+		// make room for allocated game servers, but not over the fleet replica count
+		if replicas+sumAllocated > fleet.Spec.Replicas {
+			replicas = fleet.LowerBoundReplicas(replicas - sumAllocated)
+		}
+
+		c.logger.WithField("gameserverset", active.ObjectMeta.Name).WithField("replicas", replicas).
+			Info("applying rolling update to active gameserverset")
+	}
+
+	return replicas, nil
+}
+
+// rollingUpdateRest applies the rolling update to the inactive GameServerSets
+func (c *Controller) rollingUpdateRest(fleet *stablev1alpha1.Fleet, rest []*stablev1alpha1.GameServerSet) error {
+	if len(rest) == 0 {
+		return nil
+	}
+
+	r, err := intstr.GetValueFromIntOrPercent(fleet.Spec.Strategy.RollingUpdate.MaxUnavailable, int(fleet.Spec.Replicas), true)
+	if err != nil {
+		return errors.Wrapf(err, "error calculating scaling gameserverset: %s", fleet.ObjectMeta.Name)
+	}
+	unavailable := int32(r)
+
+	for _, gsSet := range rest {
+		if gsSet.Status.Replicas > 0 && gsSet.Status.Replicas == gsSet.Spec.Replicas {
```

Comment:
Same thing here as above. If you invert the boolean check, your if statement can become much smaller with a continue in the body, and the code outside the if statement can be moved below, which makes understanding the flow easier.